mac80211/hostapd: Fix setting radio htmode when using mesh mode

xsokolikx opened the pull request:
When configuring the radio in legacy mode from luci, the htmode is not set correctly to NOHT, causing the radio in mesh mode to be set to HT40. 
Also, disable_vht parameter needs to be set when using wpa_supplicant NOHT/HT20/HT40 modes.

adrianschmutzler commented on package/network/services/hostapd/files/hostapd.sh:
```diff
@@ -765,6 +765,7 @@ wpa_supplicant_set_fixed_freq() {
 		NOHT) append network_data "disable_ht=1" "$N$T";;
 		HT20|VHT20) append network_data "disable_ht40=1" "$N$T";;
 		HT40*|VHT40*|VHT80*|VHT160*) append network_data "ht40=1" "$N$T";;
+		NOHT*|HT*) append network_data "disable_vht=1" "$N$T";;
```
Note that this is a case, so NOHT, HT20 and HT40* won't fall into this option ...

adrianschmutzler replied:
You could actually merge this case with the one following it.

adrianschmutzler commented:
Changes for different packages should be separate commits if possible. Despite, you lack a Signed-off-by in your commit message.

xsokolikx commented:
Ok, i realized the same problem and was fixing it. Also adding sign-off.

Commit:
mac80211: Fix setting radio htmode when using mesh mode

When configuring the radio in legacy mode from luci, the htmode is not set
correctly to NOHT, causing the radio in mesh mode to be set to HT40.

Signed-off-by: Enrique Rodríguez Valencia <enrique.rodriguez@galgus.net>

Commit:
hostapd: Add disable_vht when using NOHT/HT* modes

disable_vht parameter needs to be set when using wpa_supplicant NOHT/HT* modes.

Signed-off-by: Enrique Rodríguez Valencia <enrique.rodriguez@galgus.net>

xsokolikx commented:
Commit splitted in two:

- hostapd: Add disable_vht when using NOHT/HT* modes

- mac80211: Fix setting radio htmode when using mesh mode

Thanks for the feedback.

adrianschmutzler commented:
hostapd commit lacks a description (about why it is necessary).

adrianschmutzler commented on package/network/services/hostapd/files/hostapd.sh:
```diff
@@ -766,6 +766,9 @@ wpa_supplicant_set_fixed_freq() {
 		HT40*|VHT40*|VHT80*|VHT160*) append network_data "ht40=1" "$N$T";;
 	esac
+	case "$htmode" in
+		NOHT*|HT*) append network_data "disable_vht=1" "$N$T";;
```
You could actually merge this case with the one following it.

dangowrt replied:
Yes, you can use the default of the statement below on line 776:
```
*) append network_data "max_oper_chwidth=0" "$N$T";;
```
and turn that into
```
VHT20|VHT40) append network_data "max_oper_chwidth=0" "$N$T";;
*) append network_data "disable_vht=1" "$N$T";;
```

xsokolikx commented:
Added description to hostapd commit. Also, the hostapd commit has been merged with the next switch case.

jow- commented:
Are we sure that `disable_vht` is recognized by all hostapd variants? Iirc 11n support used to be optional at some point which caused hostapd to not recognize certain options.

xsokolikx commented:
The hostapd commit only affects wpa_supplicant configuration. After reviewing code from wpa-supplicant and wpa-supplicant-mini variants, disable_vht related parts exists on both:

```
#ifdef CONFIG_VHT_OVERRIDES
        if (ssid->disable_vht) {
                freq->vht_enabled = 0;
                return;
        }
#endif /* CONFIG_VHT_OVERRIDES */
```

jow- commented:
And `CONFIG_VHT_OVERRIDES` ends up being set in all cases? Iirc there's some conditional compilation going on, depending on which wireless drivers were selected at hostapd compilation time.

xsokolikx commented:
Yes, `CONFIG_VHT_OVERRIDES` is set in all cases:
```
~/package/network/services/hostapd$ grep -r CONFIG_VHT_OVERRIDES
files/wpa_supplicant-p2p.config:CONFIG_VHT_OVERRIDES=y
files/wpa_supplicant-mini.config:CONFIG_VHT_OVERRIDES=y
files/wpa_supplicant-full.config:CONFIG_VHT_OVERRIDES=y
files/wpa_supplicant-basic.config:CONFIG_VHT_OVERRIDES=y
```
Also, note that if _disable_vht_ parameter is not set, if you configure a radio using mesh on HT, it is set by default to VHT.

dangowrt reviewed (changes requested):
Please merge your new `case` statement with the existing one below, as shown in the comment.
Apart from that lgtm, let's get it ready and merge.

dangowrt commented:
cherry-picked onto master. thanks!
